Feature: add `take` compute for slices

connortsui20 opened the pull request:
Ports over the `take` implementation on primitive arrays to `vortex-compute` for `take` on generic slices.

connortsui20 committed:
add take for slices, missing avx2 implementation

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

0ax1 commented on vortex-compute/src/take/slice/portable.rs:
```diff
@@ -0,0 +1,110 @@
+///
+/// Returns a `Buffer<T>` where each element corresponds to `values[indices[i]]`.
+#[multiversion(targets("x86_64+avx2", "x86_64+avx", "aarch64+neon"))]
+fn take_portable_simd<T, I, const LANE_COUNT: usize>(values: &[T], indices: &[I]) -> Buffer<T>
```
Can you clarify what the goal of this PR is? The current implementation of `compute/take/portable/take_portable_simd` is identical.

0ax1 replied:
`take_portable_simd` and `take_primitive_scalar` take slices as input parameters already.

connortsui20 replied:
yes this is just moving them into `vortex-compute` since we're eventually going to delete all of the compute functions

0ax1 replied:
Ah right, we're just not doing it as part of this PR.

0ax1 replied:
Is it maybe better to do that move in one big PR atomically?

connortsui20 replied:
What I really should do is move that implementation here and have the primitive array call into this, but I want to wait until I add the avx2 impl plus the pvector take impl (which will deal with nullable indices)

0ax1 replied:
Yeah - However we sequence this, it'd be good to not to have duplicated code intermediately.

connortsui20 replied:
I've changed it so that the existing portable compute function on the array calls into this new one in `vortex-compute`

connortsui20 replied:
ok I've just ported the whole thing over

0ax1 commented on vortex-vector/src/primitive/macros.rs:
```diff
@@ -184,3 +220,39 @@ macro_rules! match_each_integer_pvector_mut {
+/// # Panics
+///
+/// Panics if the vector passed in to the macro is not an unsigned vector variant.
+#[macro_export]
```
Why do we introduce this twice?

connortsui20 replied:
this one is for mutable, and the only reason I also added this is for symmetry

connortsui20 committed:
call into vortex-compute instead of duplication

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

a10y commented:
Are we going to handle nullability or is that handled a level up?

connortsui20 commented:
@a10y yeah I think that it doesnt really make sense to handle nullable indices for a plain buffer, once we add take on things like vectors then we can take a primitive array or pvector as indices

joseph-isaacs commented:
Can we name this PR appropriately?

connortsui20 commented:
@joseph-isaacs why is it not named appropriately? Do you mean it should be named a migration instead?

connortsui20 commented:
ok I'm just going to port it over now to make things cleaner

cc @a10y since I believe you were the one who implemented the AVX2 impl? Everything is identical except that I pulled out the validity (which was just getting passed in so that it could construct a `PrimitiveArray`) and I changed a buffer transmute to use a new `cast_into` method I added on `Buffer`.

connortsui20 committed:
port avx2 implementation over to vortex-compute

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

a10y reviewed (approved):
lgtm, avx2 tests are running in CI and still passing
